File: tools/MobaTool.py
import base64

from Crypto.Cipher import AES
from Crypto.Hash import SHA512
from Registry import Registry

from .PrintTool import print_red, print_dict, print_yellow


class MobaXtermCipher:
    '''
    @author:        @clemthi on github 
    @repository:  https://github.com/clemthi/mobaxterm-decrypt-pass
    '''

    def __init__(self, master_password: bytes):
        self._key = SHA512.new(master_password).digest()[0:32]
        self._cipher = None

    def _generate_cipher(self) -> None:
        init_vector = AES.new(key=self._key, mode=AES.MODE_ECB).encrypt(b"\x00" * AES.block_size)
        self._cipher = AES.new(key=self._key, iv=init_vector, mode=AES.MODE_CFB, segment_size=8)

    def decrypt_password(self, cipher_text: str) -> bytes:
        self._generate_cipher()
        return self._cipher.decrypt(base64.b64decode(cipher_text)).decode('utf8')


def getEncFromIni(file: str) -> dict:
    """
    从配置文件中提取密码和凭据
    """
    print('[提示]---->从配置文件获取数据')
    with open(file, 'r', encoding='gb2312') as fr:
        data = fr.readlines()
        passwords = []
        credentials = []
        fi = 0
        ff = False
        ci = 0
        cf = False
        end = []
        ei = 0
        for v in data:
            if v == '\n':
                end.append(ei)
            if v == '[Passwords]\n':
                fi = ei
            if v == '[Credentials]\n':
                ci = ei
            ei += 1
        for i in end:
            if i > fi and not ff:
                tmp = data[fi + 1:i]
                for v in tmp:
                    passwords.append(v.strip())
                ff = True
            if i > ci and not cf and ci != 0:
                tmp = data[ci + 1:i]
                for v in tmp:
                    credentials.append(v.strip())
                cf = True
            if cf and ff:
                break
    return passwords, credentials


def getEncFromReg(file: str) -> dict:
    """
    从注册表中提取密码和凭据
    """
    passwords = []
    credentials = []
    print('[提示]---->从注册表文件获取数据')
    try:
        reg = Registry.Registry(file)
        pwdReg = reg.open("Software\\Mobatek\\MobaXterm\\P").values()
        for v in pwdReg:
            passwords.append(v.name() + '=' + v.value())
        creReg = reg.open("Software\\Mobatek\\MobaXterm\\C").values()
        for v in creReg:
            credentials.append(v.name() + '=' + v.value())
    except:
        print_red('[错误]---->注册表中未找到MobaXterm保存的密码和凭证信息，请从MobaXterm.ini配置文件中提取')
    return passwords, credentials


def analyzeMoba(file: str, masterkey: str, flag: int):
    passwords = []
    credentials = []
    cipher = MobaXtermCipher(masterkey.encode('utf8'))
    pwds = []
    cres = []
    if flag == 0:
        pwds, cres = getEncFromReg(file)
    elif flag == 1:
        pwds, cres = getEncFromIni(file)
    for pwd in pwds:
        password = {}
        tmp = pwd.split(':')
        if pwd.__contains__(':'):
            password.update({'协议与端口': tmp[0]})
        else:
            password.update({'协议与端口': ''})
        t = tmp[-1]
        index = t.index('=')
        connection = t[:index]
        encPwd = t[index + 1:]
        password.update({'连接账户与地址': connection})
        password.update({'保存的密码': cipher.decrypt_password(encPwd)})
        passwords.append(password)
    for cre in cres:
        credential = {}
        index = cre.index('=')
        name = cre[:index]
        credential.update({'凭据名': name})
        t = cre[index + 1:].split(':')
        credential.update({'用户名': t[0]})
        credential.update({'保存的密码': cipher.decrypt_password(t[1])})
        credentials.append(credential)
    dic = {}
    try:
        print_dict(passwords, passwords[0].keys(), )
        dic.update({'MobaXterm连接信息': passwords})
        print_dict(credentials, credentials[0].keys(), 'MovaXterm凭据信息')
        dic.update({'MovaXterm凭据信息': credentials})
    except:
        pass
    print_yellow('[提示]---->如果内容显示不全，请将终端全屏后再次执行')
    return dic

File: analyzers/extractor/tool/mobaTool.go
package tool

import (
	"crypto/aes"
	"crypto/sha512"
	"fmt"
	"github.com/WXjzcccc/registry"
	"github.com/deatil/go-cryptobin/cryptobin/crypto"
	"github.com/iancoleman/orderedmap"
	"golang.org/x/text/encoding/simplifiedchinese"
	"os"
	"strings"
)

func decryptMoba(ciphertext, masterPasswd string) string {
	hasher := sha512.New()
	hasher.Write([]byte(masterPasswd))
	pwd := hasher.Sum(nil)[:32]
	iv := crypto.FromHexString(strings.Repeat("00", aes.BlockSize)).WithKey(pwd).Aes().ECB().Encrypt().ToBytes()
	return crypto.FromBase64String(ciphertext).WithKey(pwd).WithIv(iv).Aes().CFB8().Decrypt().ToString()
}

func getEncFromIni(file string) ([]string, []string, error) {
	/*
		@param file	:	MobaXterm.ini配置文件路径
		@return		:	密码列表、凭据列表、错误
	*/
	content, err := os.ReadFile(file)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to read ini file: %v", err)
	}

	decoder := simplifiedchinese.GBK.NewDecoder()
	utf8Content, err := decoder.Bytes(content)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to convert encoding: %v", err)
	}

	lines := strings.Split(string(utf8Content), "\n")

	var passwords []string
	var credentials []string

	inPasswords := false
	inCredentials := false

	for _, line := range lines {
		line = strings.TrimSpace(line)
		switch line {
		case "[Passwords]":
			inPasswords = true
			inCredentials = false
			continue
		case "[Credentials]":
			inCredentials = true
			inPasswords = false
			continue
		case "":
			inPasswords = false
			inCredentials = false
			continue
		}
		if inPasswords {
			passwords = append(passwords, line)
		}
		if inCredentials {
			credentials = append(credentials, line)
		}
	}
	return passwords, credentials, nil
}

func getPwdOrCredential(reg *registry.Key, path string) ([]string, error) {
	/*
		@param reg	:	注册表实例
		@param path	:	key路径
		@return		:	提取的数据列表，错误
	*/
	var data []string
	nameKey, err := reg.OpenSubKey(path)
	if err != nil {
		return nil, fmt.Errorf("未找到保存的凭据或是密码: %v", err)
	}
	defer nameKey.Close()
	names, err := nameKey.ReadValueNames(-1)
	if err != nil {
		return nil, fmt.Errorf("failed to read registry key: %v", err)
	}
	for _, name := range names {
		value, _, err := nameKey.GetStringValue(name)
		if err != nil {
			continue
		}
		data = append(data, fmt.Sprintf("%s=%s", name, value))
	}
	return data, nil
}

func getEncFromRegistry(file string) ([]string, []string, error) {
	/*
		@param file	:	NTUSER.DAT配置文件路径
		@return		:	密码列表、凭据列表、错误
	*/
	reg, err := registry.Open(file)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to open registry file: %v", err)
	}
	defer reg.Close()
	rootKey, err := reg.OpenKey("Software\\Mobatek\\MobaXterm")
	if err != nil {
		return nil, nil, fmt.Errorf("failed to open registry key: %v", err)
	}
	defer rootKey.Close()
	passwords, _ := getPwdOrCredential(&rootKey, "P") //这俩的错误不处理
	credentials, _ := getPwdOrCredential(&rootKey, "C")
	return passwords, credentials, nil
}

func AnalyzeMobaXterm(file, masterPasswd string) (*orderedmap.OrderedMap, error) {
	/*
		@param	file:	配置文件或注册表文件
		@return:		解析结果、错误
	*/
	fileInfo, err := os.Stat(file)
	if err != nil {
		return nil, err
	}
	if fileInfo.IsDir() {
		return nil, fmt.Errorf("%s is not a file", file)
	}
	result := orderedmap.New()
	result.SetEscapeHTML(false)
	var passwords []string
	var credentials []string
	if strings.HasSuffix(file, ".ini") {
		passwords, credentials, err = getEncFromIni(file)
		if err != nil {
			return nil, err
		}
	} else {
		passwords, credentials, err = getEncFromRegistry(file)
		if err != nil {
			return nil, err
		}
	}
	var passwordList []*orderedmap.OrderedMap
	for _, password := range passwords {
		info := orderedmap.New()
		info.SetEscapeHTML(false)
		first := ""
		if strings.Contains(password, ":") {
			firstSp := strings.Split(password, ":")
			first = firstSp[0]
		}
		idx := strings.Index(password, "=")
		server := password[len(first)+1 : idx]
		pwd := decryptMoba(password[idx+1:], masterPasswd)
		info.Set("协议与端口", first)
		info.Set("连接账户与地址", server)
		info.Set("保存的密码", pwd)
		passwordList = append(passwordList, info)
	}
	var credentialList []*orderedmap.OrderedMap
	for _, credential := range credentials {
		info := orderedmap.New()
		info.SetEscapeHTML(false)
		user := ""
		encPwd := ""
		idx := strings.Index(credential, "=")
		server := credential[:idx]
		if strings.Contains(credential[idx+1:], ":") {
			user = strings.Split(credential[idx+1:], ":")[0]
			encPwd = strings.Split(credential[idx+1:], ":")[1]
		}
		pwd := decryptMoba(encPwd, masterPasswd)
		info.Set("凭据名", server)
		info.Set("用户名", user)
		info.Set("保存的密码", pwd)
		credentialList = append(credentialList, info)
	}
	result.Set("密码信息", passwordList)
	result.Set("凭据信息", credentialList)
	return result, nil
}
